canonical-web-and-design/maas-ui
# refactor(ui): refactor Formik components to be more modular

## Done

- Removed `FormCardButtons` component and updated `FormikFormButtons` to be able to handle all button types
- Reorganised the prop distribution between the Formik components. `FormikForm` is now just a convenience wrapper for `Formik` that passes all non-formik props to `FormikFormContent`. I think this allows for a cleaner separation between what's formik's functionality and what's "our" functionality.
- Similarly, `ActionForm` acts as a wrapper for `FormikForm`, setting some prop defaults and with an extra layer of functionality. I hope this new separation of props/components will make it easier to make new form components like this (e.g. `ToggleForm` for a form that can toggle between editing state).
- Updated `FormikFormButtons` prop names to make more sense in the context of using `FormikForm` (e.g. `bordered` is now `buttonsBordered`)
- Fixed types for all formik components, and the ensuing errors that came up as a result. Most of these were to do with fixing disparities between a form's `Values` type and what was given in `initialValues`, mostly things like `number | "" is not assignable to type number`. There were also a few type errors to do with the action params type not matching `Values`.

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Test that one of each "type" of form still looks and works as expected:
  - A settings form, like in http://0.0.0.0:8400/MAAS/r/settings/storage
  - A normal add form, like in http://0.0.0.0:8400/MAAS/r/settings/users/add
  - A secondary submit form, like in http://0.0.0.0:8400/MAAS/r/machines/add
  - A header action form, like in http://0.0.0.0:8400/MAAS/r/machines > click "Take action" > click "Tag"
  - An inline form like in http://0.0.0.0:8400/MAAS/r/machine/txwhnm/summary > click "able-chimp.maas"

## Fixes

Fixes #2593
